### PR TITLE
Add support for connect start/end events.

### PR DESCRIPTION
**Pull request by dave-r12**
I've altered the signatures of the connect start and end methods. Notably:

- Changed from `InetAddress` and `port` to `InetSocketAddress`. Was just easier to pull this off the route, but don't have strong feelings on this one.
- Add `Proxy` to connectStart. I thought this might be useful for people using proxies.
- Changed to use `Protocol` on connectEnd. This is already a public type and felt easier to work with.

### CODE REVIEW

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java [hunk 1]
```diff
@@ -43,18 +47,27 @@
 import static org.junit.Assert.fail;
 
 public final class EventListenerTest {
-  @Rule public final MockWebServer server = new MockWebServer();
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: Why this change? Don't see any reason in the code, unless I missed something.

Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: There is currently a dependency when using the `SocksProxy` (I added a test that uses it) to shutdown the MWS first. Alternatively, SocksProxy could be adjusted to behave like MWS by closing all open connections.

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java [hunk 2]
```diff
@@ -212,8 +228,178 @@
...
+    socksProxy = new SocksProxy();
+    socksProxy.play();
+    Proxy proxy = socksProxy.proxy();
+
+    client = client.newBuilder()
+        .proxy(proxy)
+        .build();
+
+    Call call = client.newCall(new Request.Builder()
+        .url("http://" + SocksProxy.HOSTNAME_THAT_ONLY_THE_PROXY_KNOWS + ":" + server.getPort())
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: neat

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java [hunk 3]
```diff
@@ -208,18 +217,23 @@ private void connectTunnel(int connectTimeout, int readTimeout, int writeTimeout
       rawSocket = null;
       sink = null;
       source = null;
+      eventListener.connectEnd(call, route.socketAddress(), null, null);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: this is a failure? I wonder if we can get the exception in there somehow

File: okhttp/src/main/java/okhttp3/EventListener.java [hunk 4]
```diff
@@ -61,7 +63,18 @@ public void dnsEnd(Call call, String domainName, @Nullable List<InetAddress> ine
...
+  /**
+   * Invoked just prior to initiating a socket connection.
+   *
+   * <p>This method will be invoked if no existing connection in the {@link ConnectionPool} can be
+   * reused.
+   *
+   * <p>This can be invoked more than 1 time for a single {@link Call}. For example, if the response
+   * to the {@link Call#request()} is a redirect to a different address, or a connection is retried.
+   *
+   * <p>{@code proxy} will be non-null if an HTTP or SOCKS proxy is used and null otherwise.
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/EventListener.java]: Should we use `Proxy.NO_PROXY` ? I don’t know

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java (hunk 3 above)
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: This is an oddball case: we attempted to create a tunnel, and forced to authenticate. The server closed the connection so we must reconnect with the authentication details. The connection itself is perfectly fine, but we can't derive the protocol at this point.

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 4 above)
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/EventListener.java]: Sure. Are you reluctant to include the proxy?

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/EventListener.java]: I think that having a proxy field is okay. When there is no proxy should we use `Proxy.NO_PROXY` or should we use null?

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/EventListener.java]: (elsewhere we use `null` to indicate “we haven’t selected a proxy” and `NO_PROXY` to indicate “we’ve decided to use no proxy”)

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java (hunk 3 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: aahh. Makes sense.

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 4 above)
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/EventListener.java]: Gotcha - `NO_PROXY` makes sense to me. But I'll think about it a bit more.

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java (hunk 1 above)
Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: I've made a small adjustment to `SocksProxy` so that this can continue to be a `@Rule`.

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 4 above)
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/EventListener.java]: To clarify why I added this in the first place:
I noticed that when using a SOCKS proxy, that simply passing an `InetSocketAddress` wasn't enough because the socket is transparently using the SOCKS proxy. That is why I thought adding the Proxy would be helpful.

I've changed this to be `NO_PROXY`. The intent of this (in my mind) is to communicate the proxy being used to connect. It's not intended to report the outcome of the proxy selection.

**Comment by swankjesse** [comment]
Good stuff.
